<!--
  1 failure:  :disappointed_rel...
  0 warning: 
  1 messages
  
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :disappointed_relieved: `npm test` failed. See below for details.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  ```
> titanium-mobile@7.2.0 test /Users/build/jenkins/workspace/sdk_titanium_mobile_PR-9885-6MOVYCFPCTNY7GLLJNWVQO5NN454U654ZCT2Z7A57XO24PP77C6Q
> grunt

Running "appcJs:src" (appcJs) task

Running "eslint:src" (eslint) task

/Users/build/jenkins/workspace/sdk_titanium_mobile_PR-9885-6MOVYCFPCTNY7GLLJNWVQO5NN454U654ZCT2Z7A57XO24PP77C6Q/android/cli/commands/_buildModule.js
  554:18  warning  Found non-literal argument to RegExp Constructor  security/detect-non-literal-regexp

/Users/build/jenkins/workspace/sdk_titanium_mobile_PR-9885-6MOVYCFPCTNY7GLLJNWVQO5NN454U654ZCT2Z7A57XO24PP77C6Q/tests/Resources/ti.ui.picker.addontest.js
   85:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  122:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  123:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  171:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  172:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  175:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  176:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  245:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  246:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  252:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  253:5  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions

✖ 12 problems (0 errors, 12 warnings)


Running "continue:on" task

Running "retire:js" (retire) task
Ignoring {"paths":["growl"],"descriptors":[]}
Downloading https://raw.github.com/RetireJS/retire.js/master/repository/jsrepository.json ...
(node:68194) [DEP0013] DeprecationWarning: Calling an asynchronous function without callback is deprecated.
Checking: dangerfile.js
Checking: Gruntfile.js
Checking: apidoc/docgen.js
Checking: apidoc/lib/addon_generator.js
Checking: apidoc/lib/changes_generator.js
Checking: apidoc/lib/common.js
Checking: apidoc/lib/html_generator.js
Checking: apidoc/lib/jsca_generator.js
Checking: apidoc/lib/jsduck_generator.js
Checking: apidoc/lib/json_generator.js
Checking: apidoc/lib/modulehtml_generator.js
Checking: apidoc/lib/parity_generator.js
Checking: apidoc/lib/solr_generator.js
Checking: apidoc/validate.js
Checking: build/android.js
Checking: build/androidsdk.js
Checking: build/ant.js
Checking: build/docs.js
Checking: build/git.js
Checking: build/ios.js
Checking: build/packager.js
Checking: build/scons-build.js
Checking: build/scons-clean.js
Checking: build/scons-cleanbuild.js
Checking: build/scons-install.js
Checking: build/scons-modules-integrity.js
Checking: build/scons-package.js
Checking: build/scons-ssri.js
Checking: build/scons-test.js
Checking: build/scons-update-node-deps.js
Checking: build/scons.js
Checking: build/titanium-sdk/lib/titanium.js
Checking: build/utils.js
Checking: build/windows.js
Checking: cli/commands/build.js
Checking: cli/commands/clean.js
Checking: cli/commands/create.js
Checking: cli/commands/project.js
Checking: cli/hooks/hyperloop-fix.js
Checking: cli/lib/creator.js
Checking: cli/lib/creators/app.js
Checking: cli/lib/creators/applewatch.js
Checking: cli/lib/creators/module.js
Checking: cli/lib/info.js
Checking: android/cli/commands/_build.js
Checking: android/cli/commands/_buildModule.js
Checking: android/cli/hooks/aar-transform.js
Checking: android/cli/hooks/package.js
Checking: android/cli/hooks/run.js
Checking: android/cli/lib/AndroidManifest.js
Checking: android/cli/lib/base-builder.js
Checking: android/cli/lib/detect.js
Checking: android/cli/lib/info.js
Checking: android/cli/tests/run.js
Checking: android/cli/tests/test-androidmanifest.js
Checking: iphone/cli/commands/_build.js
Checking: iphone/cli/commands/_buildModule.js
Checking: iphone/cli/hooks/frameworks.js
Checking: iphone/cli/hooks/hyperloop.js
Checking: iphone/cli/hooks/install.js
Checking: iphone/cli/hooks/package.js
Checking: iphone/cli/hooks/run.js
Checking: iphone/cli/lib/info.js
Checking: tests/Resources/ti.api.addontest.js
Checking: tests/Resources/ti.example.addontest.js
Checking: tests/Resources/ti.network.httpclient.addontest.js
Checking: tests/Resources/ti.ui.picker.addontest.js
Checking: tests/Resources/ti.ui.tableview.addontest.js
Checking: tests/Resources/ti.utils.test.js
No vulnerabilities found.

Running "retire:node" (retire) task
Ignoring {"paths":["growl"],"descriptors":[]}
Downloading https://raw.github.com/RetireJS/retire.js/master/repository/npmrepository.json ...
(node:68194) [DEP0013] DeprecationWarning: Calling an asynchronous function without callback is deprecated.
Checking: .
Missing version for babel-eslint. Need to run npm install ?
Missing version for eslint-plugin-jsx-a11y. Need to run npm install ?
Missing version for eslint-plugin-react. Need to run npm install ?
Missing version for babel-eslint. Need to run npm install ?
Missing version for eslint-plugin-jsx-a11y. Need to run npm install ?
Missing version for eslint-plugin-react. Need to run npm install ?
No vulnerabilities found.

Running "continue:off" task

Running "ios_format:src" (ios_format) task

Running "android_format:src" (android_format) task
Fatal error: Error: Formatting incorrect on "android/titanium/src/java/org/appcelerator/titanium/TiBlob.java", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='9868' length='0'> </replacement>
<replacement offset='10162' length='4'>&#10;				</replacement>
<replacement offset='10171' length='5'>&#10;					</replacement>
<replacement offset='10196' length='0'> </replacement>
<replacement offset='10219' length='4'>&#10;				</replacement>
<replacement offset='10248' length='5'>&#10;					</replacement>
<replacement offset='10283' length='5'>&#10;					</replacement>
<replacement offset='10300' length='4'>&#10;				</replacement>
</replacements>

npm ERR! Test failed.  See above for more details.

```
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

--- a/android/modules/network/src/java/ti/modules/titanium/network/TiHTTPClient.java
+++ b/android/modules/network/src/java/ti/modules/titanium/network/TiHTTPClient.java
@@ -895,14 +895,7 @@ public class TiHTTPClient
 				String mimeType = blob.getMimeType();
 				File tmpFile =
 					File.createTempFile("tixhr", "." + TiMimeTypeHelper.getFileExtensionFromMimeType(mimeType, "txt"));
-				if (blob.getType() == TiBlob.TYPE_STREAM_BASE64) {
-					FileOutputStream fos = new FileOutputStream(tmpFile);
-					TiBaseFile.copyStream(blob.getInputStream(),
-										  new Base64OutputStream(fos, android.util.Base64.DEFAULT));
-					fos.close();
-				} else {
-					createFileFromBlob(blob, tmpFile);
-				}
+				createFileFromBlob(blob, tmpFile);
 
 				tmpFiles.add(tmpFile);
 

--- a/android/modules/utils/src/java/ti/modules/titanium/utils/UtilsModule.java
+++ b/android/modules/utils/src/java/ti/modules/titanium/utils/UtilsModule.java
@@ -34,12 +34,17 @@ public class UtilsModule extends KrollModule
 		super();
 	}
 
-	private String convertToString(Object obj)
+	private byte[] convertToBytes(Object obj)
 	{
 		if (obj instanceof String) {
-			return (String) obj;
+			try {
+				return ((String) obj).getBytes("UTF-8");
+			} catch (UnsupportedEncodingException e) {
+				Log.e(TAG, "UTF-8 is not a supported encoding type");
+			}
+			return ((String) obj).getBytes(); // should never fall back here!
 		} else if (obj instanceof TiBlob) {
-			return ((TiBlob) obj).getText();
+			return ((TiBlob) obj).getBytes();
 		} else {
 			throw new IllegalArgumentException("Invalid type for argument");
 		}
@@ -48,22 +53,14 @@ public class UtilsModule extends KrollModule
 	@Kroll.method
 	public TiBlob base64encode(Object obj)
 	{
-		if (obj instanceof TiBlob) {
-			return TiBlob.blobFromString(((TiBlob) obj).toBase64());
-		} else if (obj instanceof TiFileProxy) {
-			try {
-				return TiBlob.blobFromStreamBase64(
-					((TiFileProxy) obj).getInputStream(),
-					TiMimeTypeHelper.getMimeType(((TiFileProxy) obj).getBaseFile().nativePath()));
-			} catch (IOException e) {
-				Log.e(TAG, "Problem reading file");
-			}
+		if (obj instanceof TiFileProxy) {
+			// recursively call base64encode after converting Ti.Filesystem.File to a Ti.Blob wrapping it
+			return base64encode(TiBlob.blobFromFile(((TiFileProxy) obj).getBaseFile()));
 		}
-		String data = convertToString(obj);
+		byte[] data = convertToBytes(obj);
 		if (data != null) {
 			try {
-				return TiBlob.blobFromString(
-					new String(Base64.encode(data.getBytes("UTF-8"), Base64.NO_WRAP), "UTF-8"));
+				return TiBlob.blobFromString(new String(Base64.encode(data, Base64.NO_WRAP), "UTF-8"));
 			} catch (UnsupportedEncodingException e) {
 				Log.e(TAG, "UTF-8 is not a supported encoding type");
 			}
@@ -74,13 +71,13 @@ public class UtilsModule extends KrollModule
 	@Kroll.method
 	public TiBlob base64decode(Object obj)
 	{
-		String data = convertToString(obj);
+		if (obj instanceof TiFileProxy) {
+			// recursively call base64decode after converting Ti.Filesystem.File to a Ti.Blob wrapping it
+			return base64decode(TiBlob.blobFromFile(((TiFileProxy) obj).getBaseFile()));
+		}
+		byte[] data = convertToBytes(obj);
 		if (data != null) {
-			try {
-				return TiBlob.blobFromData(Base64.decode(data.getBytes("UTF-8"), Base64.NO_WRAP));
-			} catch (UnsupportedEncodingException e) {
-				Log.e(TAG, "UTF-8 is not a supported encoding type");
-			}
+			return TiBlob.blobFromData(Base64.decode(data, Base64.NO_WRAP));
 		}
 		return null;
 	}
@@ -88,10 +85,7 @@ public class UtilsModule extends KrollModule
 	@Kroll.method
 	public String md5HexDigest(Object obj)
 	{
-		if (obj instanceof TiBlob) {
-			return DigestUtils.md5Hex(((TiBlob) obj).getBytes());
-		}
-		String data = convertToString(obj);
+		byte[] data = convertToBytes(obj);
 		if (data != null) {
 			return DigestUtils.md5Hex(data);
 		}
@@ -101,10 +95,7 @@ public class UtilsModule extends KrollModule
 	@Kroll.method
 	public String sha1(Object obj)
 	{
-		if (obj instanceof TiBlob) {
-			return DigestUtils.shaHex(((TiBlob) obj).getBytes());
-		}
-		String data = convertToString(obj);
+		byte[] data = convertToBytes(obj);
 		if (data != null) {
 			return DigestUtils.shaHex(data);
 		}
@@ -123,13 +114,7 @@ public class UtilsModule extends KrollModule
 		// NOTE: DigestUtils with the version before 1.4 doesn't have the function sha256Hex,
 		// so we deal with it ourselves
 		try {
-			byte[] b = null;
-			if (obj instanceof TiBlob) {
-				b = ((TiBlob) obj).getBytes();
-			} else {
-				String data = convertToString(obj);
-				b = data.getBytes();
-			}
+			byte[] b = convertToBytes(obj);
 			MessageDigest algorithm = MessageDigest.getInstance("SHA-256");
 			algorithm.reset();
 			algorithm.update(b);
@@ -141,33 +126,6 @@ public class UtilsModule extends KrollModule
 			return result.toString();
 		} catch (NoSuchAlgorithmException e) {
 			Log.e(TAG, "SHA256 is not a supported algorithm");
-		}
-		return null;
-	}
-
-	public String transcodeString(String orig, String inEncoding, String outEncoding)
-	{
-		try {
-
-			Charset charsetOut = Charset.forName(outEncoding);
-			Charset charsetIn = Charset.forName(inEncoding);
-
-			ByteBuffer bufferIn = ByteBuffer.wrap(orig.getBytes(charsetIn.name()));
-			CharBuffer dataIn = charsetIn.decode(bufferIn);
-			bufferIn.clear();
-			bufferIn = null;
-
-			ByteBuffer bufferOut = charsetOut.encode(dataIn);
-			dataIn.clear();
-			dataIn = null;
-			byte[] dataOut = bufferOut.array();
-			bufferOut.clear();
-			bufferOut = null;
-
-			return new String(dataOut, charsetOut.name());
-
-		} catch (UnsupportedEncodingException e) {
-			Log.e(TAG, "Unsupported encoding: " + e.getMessage(), e);
 		}
 		return null;
 	}

--- a/android/titanium/src/java/org/appcelerator/titanium/TiBlob.java
+++ b/android/titanium/src/java/org/appcelerator/titanium/TiBlob.java
@@ -65,12 +65,6 @@ public class TiBlob extends KrollProxy
 	 */
 	public static final int TYPE_STRING = 3;
 
-	/**
-	 * Represents a Blob that contains stream data that needs to be converted to base64.
-	 * @module.api
-	 */
-	public static final int TYPE_STREAM_BASE64 = 4;
-
 	private int type;
 	private Object data;
 	private String mimetype;
@@ -111,16 +105,6 @@ public class TiBlob extends KrollProxy
 	public static TiBlob blobFromFile(TiBaseFile file)
 	{
 		return blobFromFile(file, TiMimeTypeHelper.getMimeType(file.nativePath()));
-	}
-
-	/**
-	 * Creates a blob from a stream to convert to base64.
-	 * @param stream the stream used to create blob.
-	 * @return new instane of TiBlob.
-	 */
-	public static TiBlob blobFromStreamBase64(InputStream stream, String mimeType)
-	{
-		return new TiBlob(TYPE_STREAM_BASE64, stream, mimeType);
 	}
 
 	/**
@@ -209,7 +193,10 @@ public class TiBlob extends KrollProxy
 	{
 		String mt = null;
 		InputStream is = getInputStream();
-		if (is != null) {
+		// We shouldn't try and sniff content type if mark isn't supported by this
+		// input stream! Otherwise we'll read bytes that we can't stuff back anymore
+		// so the stream will have been modified for future reads.
+		if (is != null && is.markSupported()) {
 			try {
 				mt = URLConnection.guessContentTypeFromStream(is);
 				if (mt == null) {
@@ -234,13 +221,16 @@ public class TiBlob extends KrollProxy
 		if (is != null) {
 			try {
 
-				// Look ahead up to 64 bytes for the longest encoded header
-				is.mark(64);
-				byte[] bytes = new byte[64];
+				// Look ahead up to 12 bytes (highest number of bytes we care about for now)
+				is.mark(12);
+				byte[] bytes = new byte[12];
 				int length = is.read(bytes);
+				is.reset();
 				if (length == -1) {
 					return null;
 				}
+
+				// This is basically exactly what the normal JDK sniffs for, but Android's fork does not
 				if (bytes[0] == 'G' && bytes[1] == 'I' && bytes[2] == 'F' && bytes[3] == '8') {
 					mt = "image/gif";
 				} else if (bytes[0] == (byte) 0x89 && bytes[1] == (byte) 0x50 && bytes[2] == (byte) 0x4E
@@ -335,20 +325,6 @@ public class TiBlob extends KrollProxy
 					}
 				}
 				break;
-			case TYPE_STREAM_BASE64:
-				InputStream inStream = (InputStream) data;
-				if (inStream != null) {
-					try {
-						bytes = KrollStreamHelper.toByteArray(inStream, getLength());
-					} finally {
-						try {
-							inStream.close();
-						} catch (IOException e) {
-							Log.w(TAG, e.getMessage(), e);
-						}
-					}
-				}
-				break;
 			default:
 				throw new IllegalArgumentException("Unknown Blob type id " + type);
 		}
@@ -374,8 +350,6 @@ public class TiBlob extends KrollProxy
 			case TYPE_DATA:
 			case TYPE_IMAGE:
 				return ((byte[]) data).length;
-			case TYPE_STREAM_BASE64:
-				throw new IllegalStateException("Not yet implemented. TYPE_STREAM_BASE64");
 			default:
 				// this is probably overly expensive.. is there a better way?
 				return getBytes().length;
@@ -396,8 +370,6 @@ public class TiBlob extends KrollProxy
 					Log.e(TAG, e.getMessage(), e);
 					return null;
 				}
-			case TYPE_STREAM_BASE64:
-				return (InputStream) data;
 			default:
 				return new ByteArrayInputStream(getBytes());
 		}
@@ -427,8 +399,6 @@ public class TiBlob extends KrollProxy
 				break;
 			case TYPE_FILE:
 				throw new IllegalStateException("Not yet implemented. TYPE_FILE");
-			case TYPE_STREAM_BASE64:
-				throw new IllegalStateException("Not yet implemented. TYPE_STREAM_BASE64");
 				// break;
 			default:
 				throw new IllegalArgumentException("Unknown Blob type id " + type);
@@ -463,8 +433,6 @@ public class TiBlob extends KrollProxy
 					Log.w(TAG, "Unable to convert to string.");
 				}
 				break;
-			case TYPE_STREAM_BASE64:
-				throw new IllegalStateException("Not yet implemented. TYPE_STREAM_BASE64");
 		}
 
 		return result;

--- a/build/scons-test.js
+++ b/build/scons-test.js
@@ -17,6 +17,7 @@ program
 	.option('-C, --device-id [id]', 'Titanium device id to run the unit tests on. Only valid when there is a target provided')
 	.option('-T, --target [target]', 'Titanium platform target to run the unit tests on. Only valid when there is a single platform provided')
 	.option('-s, --skip-sdk-install', 'Skip the SDK installation step')
+	.option('-b, --branch [branch]', 'Which branch of the test suite to use', 'master')
 	.parse(process.argv);
 
 let platforms = program.args;
@@ -44,9 +45,10 @@ if (!program.skipSdkInstall && !fs.existsSync(zipfile)) {
  * SDK zipfile in dist against the supplied platforms.
  *
  * @param  {String[]}   platforms [description]
+ * @param  {String}   branch [description]
  * @param  {Function} next      [description]
  */
-function runTests(platforms, next) {
+function runTests(platforms, branch, next) {
 	async.series([
 		function (cb) {
 			// If we have a clone of the tests locally, wipe it...
@@ -56,7 +58,7 @@ function runTests(platforms, next) {
 			}
 			// clone the common test suite shallow
 			// FIXME Determine the correct branch of the suite to clone like we do in the Jenkinsfile
-			exec('git clone --depth 1 https://github.com/appcelerator/titanium-mobile-mocha-suite.git', { cwd: path.join(__dirname, '..') }, cb);
+			exec('git clone --depth 1 https://github.com/appcelerator/titanium-mobile-mocha-suite.git -b ' + branch, { cwd: path.join(__dirname, '..') }, cb);
 		},
 		function (cb) {
 			// Make sure it's dependencies are installed
@@ -88,7 +90,7 @@ function runTests(platforms, next) {
 	], next);
 }
 
-runTests(platforms, function (err) {
+runTests(platforms, program.branch, function (err) {
 	if (err) {
 		console.error(err.toString());
 		process.exit(1);

--- a/iphone/Classes/UtilsModule.m
+++ b/iphone/Classes/UtilsModule.m
@@ -27,6 +27,17 @@
   THROW_INVALID_ARG(@"Invalid type");
 }
 
+- (NSData *)convertToData:(id)arg
+{
+  if ([arg isKindOfClass:[TiBlob class]]) {
+    return [(TiBlob *)arg data];
+  } else if ([arg isKindOfClass:[TiFile class]]) {
+    // Support TiFile with possibly binary data by converting to TiBlob and recursing
+    return [self convertToData:[(TiFile *)arg blob]];
+  }
+  return [[self convertToString:arg] dataUsingEncoding:NSUTF8StringEncoding];
+}
+
 - (NSString *)apiName
 {
   return @"Ti.Utils";
@@ -37,16 +48,8 @@
 - (TiBlob *)base64encode:(id)args
 {
   ENSURE_SINGLE_ARG(args, NSObject);
-  NSData *data;
-
-  if ([args isKindOfClass:[TiBlob class]]) {
-    data = [(TiBlob *)args data];
-  } else {
-    data = [[self convertToString:args] dataUsingEncoding:NSUTF8StringEncoding];
-  }
-
+  NSData *data = [self convertToData:args];
   NSString *base64Encoded = [data base64EncodedStringWithOptions:0];
-
   if (base64Encoded != nil) {
     return [[[TiBlob alloc] _initWithPageContext:[self pageContext]
                                          andData:[base64Encoded dataUsingEncoding:NSUTF8StringEncoding]
@@ -76,34 +79,29 @@
 {
   ENSURE_SINGLE_ARG(args, NSObject);
 
-  NSData *data = nil;
-  NSString *nstr = [self convertToString:args];
-  if (nstr) {
-    const char *s = [nstr UTF8String];
-    data = [NSData dataWithBytes:s length:strlen(s)];
-  } else if ([args respondsToSelector:@selector(data)]) {
-    data = [args data];
-  }
+  NSData *data = [self convertToData:args];
   return [TiUtils md5:data];
 }
 
 - (id)sha1:(id)args
 {
   ENSURE_SINGLE_ARG(args, NSObject);
-  NSString *nstr = [self convertToString:args];
-  const char *cStr = [nstr UTF8String];
+
+  NSData *data = [self convertToData:args];
+
   unsigned char result[CC_SHA1_DIGEST_LENGTH];
-  CC_SHA1(cStr, (CC_LONG)[nstr lengthOfBytesUsingEncoding:NSUTF8StringEncoding], result);
+  CC_SHA1([data bytes], [data length], result);
   return [TiUtils convertToHex:(unsigned char *)&result length:CC_SHA1_DIGEST_LENGTH];
 }
 
 - (id)sha256:(id)args
 {
   ENSURE_SINGLE_ARG(args, NSObject);
-  NSString *nstr = [self convertToString:args];
-  const char *cStr = [nstr UTF8String];
+
+  NSData *data = [self convertToData:args];
+
   unsigned char result[CC_SHA256_DIGEST_LENGTH];
-  CC_SHA256(cStr, (CC_LONG)[nstr lengthOfBytesUsingEncoding:NSUTF8StringEncoding], result);
+  CC_SHA256([data bytes], [data length], result);
   return [TiUtils convertToHex:(unsigned char *)&result length:CC_SHA256_DIGEST_LENGTH];
 }
 

--- a/tests/Resources/ti.utils.test.js
+++ b/tests/Resources/ti.utils.test.js
@@ -1,0 +1,249 @@
+/*
+ * Appcelerator Titanium Mobile
+ * Copyright (c) 2015-Present by Appcelerator, Inc. All Rights Reserved.
+ * Licensed under the terms of the Apache Public License
+ * Please see the LICENSE included with this distribution for details.
+ */
+/* eslint-env mocha */
+/* global Ti */
+/* eslint no-unused-expressions: "off" */
+'use strict';
+var should = require('./utilities/assertions'),
+	utilities = require('./utilities/utilities');
+
+// handy tool: https://www.fileformat.info/tool/hash.htm
+
+describe('Titanium.Utils', function () {
+	var win;
+
+	afterEach(function () {
+		if (win) {
+			win.close();
+		}
+		win = null;
+	});
+
+	it('exists', function () {
+		should(Ti.Utils).not.be.undefined;
+		should(Ti.Utils).be.an.Object;
+	});
+
+	it('apiName', function () {
+		should(Ti.Utils).have.readOnlyProperty('apiName').which.is.a.String;
+		should(Ti.Utils.apiName).be.eql('Ti.Utils');
+	});
+
+	it('#base64decode(String)', function () {
+		var test;
+		// Basic tests
+		should(Ti.Utils.base64decode).be.a.Function;
+		test = Ti.Utils.base64decode('dGVzdA==');
+		should(test).be.a.Object;
+		should(test.apiName).eql('Ti.Blob');
+		should(test.getText()).be.eql('test');
+
+		// Test string without padding
+		should(!Ti.Utils.base64decode('eyJzdWIiOiJ0ZXN0IiwiZW1haWwiOiJ0ZXN0IiwiYXVkIjoidGVzdCIsImp0aSI6ImxvTHM4d2o5aWxBQUtWckNxbzhaMFMiLCJpc3MiOiJodHRwczpcL1wvc3NvLmV4YW1wbGUuY29tIiwiaWF0IjoxNTI2MTY3NDc3LCJleHAiOjE1MjYxNjc0NzcsInBpLnRlc3QiOiJMSTRmMW81Q2pqU2tHU2xTanM0bHlPeVlROCJ9')).not.be.null;
+
+		// More padding tests
+		should(Ti.Utils.base64decode('Zg').text).eql('f');
+		should(Ti.Utils.base64decode('Zm8').text).eql('fo');
+		should(Ti.Utils.base64decode('Zm9v').text).eql('foo');
+		should(Ti.Utils.base64decode('Zm9vYg').text).eql('foob');
+		should(Ti.Utils.base64decode('Zm9vYmE').text).eql('fooba');
+		should(Ti.Utils.base64decode('Zm9vYmFy').text).eql('foobar');
+	});
+
+	// FIXME Windows gives: 'base64decode: attempt to decode a value not in base64 char set'
+	it.windowsBroken('#base64decode(Ti.Blob with text data)', function () {
+		var f = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, 'txtFiles/encodedFile.txt'),
+			blob = Ti.Utils.base64decode(f.read());
+		should(blob.toString()).eql('Decoding successful!');
+	});
+
+	it('#base64encode(String)', function () {
+		var test;
+		should(Ti.Utils.base64encode).be.a.Function;
+		test = Ti.Utils.base64encode('test');
+		should(test).be.a.Object;
+		should(test.apiName).eql('Ti.Blob');
+		should(test.getText()).be.eql('dGVzdA==');
+	});
+
+	it('#base64encode(Ti.Blob#TYPE_FILE)', function () {
+		var f = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, 'txtFiles/decodedFile.txt'),
+			contents = f.read();
+		should(Ti.Utils.base64encode(contents).toString()).eql('SSBhbSBub3QgZW5jb2RlZCB5ZXQu');
+	});
+
+	it('#base64encode(Ti.Blob#TYPE_DATA from Ti.UI.View.toImage() async)', function (finish) {
+		var win,
+			label;
+		this.timeout(5000);
+		win = Ti.UI.createWindow();
+		label = Ti.UI.createLabel({ text: 'test' });
+		win.add(label);
+		win.addEventListener('focus', function () {
+			// Get a Ti.Blob from view.toImage()
+			label.toImage(function (blob) {
+				var result;
+				try {
+					if (utilities.isAndroid()) {
+						should(blob.type).eql(2); // Android-specific property, value of 2 indicates TYPE_DATA
+					}
+					result = Ti.Utils.base64encode(blob);
+					// result here is a Ti.Blob
+					should(result).be.a.Object; // Fails here
+					should(result.apiName).eql('Ti.Blob');
+					// should(blob.text).eql('aGVsbG8gd29ybGQ='); // FIXME What sanity check can we do here?
+					finish();
+				} catch (err) {
+					finish(err);
+				}
+			});
+		});
+		win.open();
+	});
+
+	it('#base64encode(Ti.Blob#TYPE_DATA from Ti.Buffer.toBlob())', function () {
+		var blob,
+			buffer = Ti.createBuffer({ value: 'hello world' }); // Easiest way to get a TYPE_DATA Blob is from Ti.Buffer.toBlob()
+		blob = Ti.Utils.base64encode(buffer.toBlob());
+		// result here is a Ti.Blob
+		should(blob).be.a.Object;
+		should(blob.apiName).eql('Ti.Blob');
+		should(blob.text).eql('aGVsbG8gd29ybGQ=');
+	});
+
+	it('#base64encode(Ti.Blob#TYPE_STRING)', function () {
+		// Only way to get a blob of type string is the result of base64encode(String) on Android!
+		var blob,
+			test = Ti.Utils.base64encode('test');
+		should(test).be.a.Object;
+		should(test.apiName).eql('Ti.Blob');
+		if (utilities.isAndroid()) {
+			should(test.type).eql(3); // Android-specific property, value of 3 indicates TYPE_STRING
+		}
+		should(test.getText()).be.eql('dGVzdA==');
+
+		blob = Ti.Utils.base64encode(test);
+		// result here is a Ti.Blob
+		should(blob).be.a.Object;
+		should(blob.apiName).eql('Ti.Blob');
+		should(blob.text).eql('ZEdWemRBPT0=');
+	});
+
+	// FIXME: base64encode accepts Ti.File as a parameter on iOS/Android, but not on Windows.
+	it.windowsBroken('#base64encode(Ti.Filesystem.File)', function () {
+		var f = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, 'txtFiles/decodedFile.txt');
+		var blob = Ti.Utils.base64encode(f);
+
+		// result here is a Ti.Blob
+		should(blob).be.a.Object;
+		should(blob.apiName).eql('Ti.Blob');
+		should(blob.text).eql('SSBhbSBub3QgZW5jb2RlZCB5ZXQu');
+	});
+
+	// FIXME: base64decode accepts Ti.File as a parameter on iOS/Android, but not on Windows.
+	it.windowsBroken('#base64decode(Ti.Filesystem.File with text data)', function () {
+		var f = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, 'txtFiles/encodedFile.txt');
+		var blob = Ti.Utils.base64decode(f);
+
+		// result here is a Ti.Blob
+		should(blob).be.a.Object;
+		should(blob.apiName).eql('Ti.Blob');
+		should(blob.toString()).eql('Decoding successful!');
+	});
+
+	// FIXME: How can I make this valid? The input needs to be valid base64...
+	// An image can't be right. Maybe we can validate in UtilsModule that a given blob is non-binary?
+	// it.windowsBroken('#base64decode(Ti.Filesystem.File with binary data)', function () {
+	// 	var binaryFile = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, 'Logo.png'),
+	// 		blob = Ti.Utils.base64decode(binaryFile);
+	//
+	// 	// result here is a Ti.Blob
+	// 	should(blob).be.a.Object;
+	// 	should(blob.apiName).eql('Ti.Blob');
+	// 	// ignore the actual decoded value...
+	// });
+
+	it('#md5HexDigest(String)', function () {
+		var test;
+		should(Ti.Utils.md5HexDigest).be.a.Function;
+		test = Ti.Utils.md5HexDigest('test');
+		should(test).be.a.String;
+		should(test).be.eql('098f6bcd4621d373cade4e832627b4f6');
+	});
+
+	// FIXME Windows gives different md5 hash! Maybe line ending difference?
+	it.windowsBroken('#md5HexDigest(Ti.Blob with text data)', function () {
+		var f = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, 'txtFiles/file.txt'),
+			contents = f.read(),
+			test;
+		should(Ti.Utils.md5HexDigest).be.a.Function;
+		test = Ti.Utils.md5HexDigest(contents);
+		should(test).be.a.String;
+		should(test).be.eql('4fe8a693c64f93f65c5faf42dc49ab23'); // Windows Desktop gives: 'ab1600f840b927f80a3dc000c510d1d3'
+	});
+
+	it.windowsBroken('#md5HexDigest(Ti.Blob with binary data)', function () {
+		var binaryFile = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, 'Logo.png'),
+			blob = binaryFile.read(),
+			result;
+		should(Ti.Utils.md5HexDigest).be.a.Function;
+		result = Ti.Utils.md5HexDigest(blob);
+		should(result).be.a.String;
+		should(result).be.eql('803fd0b8dd9a3ca5238390732db54062');
+	});
+
+	it('#sha1(String)', function () {
+		var test;
+		should(Ti.Utils.sha1).be.a.Function;
+		test = Ti.Utils.sha1('test');
+		should(test).be.a.String;
+		should(test).be.eql('a94a8fe5ccb19ba61c4c0873d391e987982fbbd3');
+	});
+
+	it('#sha1(Ti.Blob with text data)', function () {
+		var textFile = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, 'txtFiles/decodedFile.txt'),
+			blob = textFile.read();
+		should(Ti.Utils.sha1(blob)).eql('ddbb50fb5beea93d1d4913fc22355c84f22d43ed');
+	});
+
+	it('#sha1(Ti.Blob with binary data)', function () {
+		var binaryFile = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, 'Logo.png'),
+			blob = binaryFile.read();
+		should(Ti.Utils.sha1(blob)).eql('668e98c66d8a11ef38ab442d9d6d4a21d8593645');
+	});
+
+	it('#sha256(String)', function () {
+		var test;
+		should(Ti.Utils.sha256).be.a.Function;
+		test = Ti.Utils.sha256('test');
+		should(test).be.a.String;
+		should(test).be.eql('9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08');
+	});
+
+	it('#sha256(Ti.Blob with text data)', function () {
+		var textFile = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, 'txtFiles/decodedFile.txt'),
+			blob = textFile.read();
+		should(Ti.Utils.sha256(blob)).eql('9f81cd4f510080f1da92386b391cf2539b21f6363df491b89787e50fbc33b2c3');
+	});
+
+	it('#sha256(Ti.Blob with binary data)', function () {
+		var binaryFile = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, 'Logo.png'),
+			blob = binaryFile.read();
+		should(Ti.Utils.sha256(blob)).eql('54be80ae48e4242d56170248e730ffac60a2828d07260a048e2ac0fd62386234');
+	});
+
+	// FIXME Android and iOS do no newlines for longer output, Windows does. Need to get parity
+	it.windowsBroken('TIMOB-25513', function () {
+		var shortString = 'ABCDEFGHIJ1234567890ABCDEFGHIJ12|psndemo2|abcd:1',
+			longString = 'ABCDEFGHIJ1234567890ABCDEFGHIJ12|psndemo2|abcd:12345678901234567890',
+			tiBase64ShortResult = Ti.Utils.base64encode(shortString),
+			tiBase64LongResult  = Ti.Utils.base64encode(longString);
+
+		should(tiBase64ShortResult.getText()).be.eql('QUJDREVGR0hJSjEyMzQ1Njc4OTBBQkNERUZHSElKMTJ8cHNuZGVtbzJ8YWJjZDox');
+		should(tiBase64LongResult.getText()).be.eql('QUJDREVGR0hJSjEyMzQ1Njc4OTBBQkNERUZHSElKMTJ8cHNuZGVtbzJ8YWJjZDoxMjM0NTY3ODkwMTIzNDU2Nzg5MA==');
+	});
+});

